File: backend/app/api/routes/tickers.py
"""Ticker search route backed by the SEC company directory."""

from typing import Annotated

from fastapi import APIRouter, Query
from pydantic import BaseModel

from app.services import tickers

router = APIRouter(prefix="/tickers", tags=["tickers"])


class Company(BaseModel):
    """A ticker search result."""

    ticker: str
    title: str
    cik: str


@router.get("/search", response_model=list[Company])
async def search_tickers(
    q: Annotated[str, Query(description="Company name or ticker fragment.")],
    limit: Annotated[int, Query(ge=1, le=50)] = 20,
) -> list[dict[str, str]]:
    """Return companies whose ticker or name matches ``q``, best matches first."""
    return await tickers.search(q, limit=limit)

File: frontend/lib/tickers.ts
const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

export type Company = {
  ticker: string;
  title: string;
  cik: string;
};

/**
 * Search the backend ticker directory (SEC-backed) by company name or ticker.
 * Pass an AbortSignal to cancel in-flight requests as the user keeps typing.
 */
export async function searchTickers(
  query: string,
  signal?: AbortSignal,
): Promise<Company[]> {
  const trimmed = query.trim();
  if (!trimmed) return [];

  const url = new URL("/tickers/search", API_BASE_URL);
  url.searchParams.set("q", trimmed);
  url.searchParams.set("limit", "20");

  const res = await fetch(url, { signal });
  if (!res.ok) {
    throw new Error(`Ticker search failed: ${res.status}`);
  }
  return (await res.json()) as Company[];
}
